WhateverGreen/kern_igfx_clock.cpp

```diff
@@ -0,0 +1,881 @@
+
+	// Call the original ReadAUX() function to read from DPCD
+	IOReturn retVal = callbackIGFX->modDPCDMaxLinkRateFix.orgReadAUX(address, buffer, length);
+
+	// Guard: Check the DPCD register address
+	// The first 16 fields of the receiver capabilities reside at 0x0 (DPCD Register Address)
+	if (address != DPCD_DEFAULT_RECEIVER_CAPS_ADDRESS && address != DPCD_EXTENDED_RECEIVER_CAPS_ADDRESS)
+		return retVal;
+
+	// The driver tries to read the first 16 bytes from DPCD (0x0000) or extended DPCD (0x2200)
+	// Get the current framebuffer index (An UInt32 field at 0x1dc in a framebuffer instance)
+	// We read the value of "IOFBDependentIndex" instead of accessing that field directly
+	uint32_t index;
+	// Guard: Should be able to retrieve the index from the registry
+	if (!callbackIGFX->modDPCDMaxLinkRateFix.getFramebufferIndex(index)) {
+		SYSLOG("igfx", "MLR: [COMM] wrapReadAUX() Failed to read the current framebuffer index.");
+		return retVal;
+	}
+
+	// Guard: Check the framebuffer index
+	// By default, FB 0 refers to the builtin display
+	if (index != 0)
+		// The driver is reading DPCD for an external display
+		return retVal;
+
+	// The driver tries to read the receiver capabilities for the builtin display
+	auto caps = reinterpret_cast<DPCDCap16*>(buffer);
+
+	// Set the custom maximum link rate value if user has specified one
+	if (callbackIGFX->modDPCDMaxLinkRateFix.maxLinkRate != 0) {
+		DBGLOG("igfx", "MLR: [COMM] wrapReadAUX() Will use the maximum link rate specified by user or cached by the previous probe call.");
+		caps->maxLinkRate = callbackIGFX->modDPCDMaxLinkRateFix.maxLinkRate;
+	} else {
+		DBGLOG("igfx", "MLR: [COMM] wrapReadAUX() Will probe the maximum link rate from the table.");
+		caps->maxLinkRate = callbackIGFX->modDPCDMaxLinkRateFix.probeMaxLinkRate();
+		// The graphics driver tries to read the maximum link rate from both DPCD address 0x0 and 0x2200.
+		// We save the probe result to avoid duplicated computation.
+		callbackIGFX->modDPCDMaxLinkRateFix.maxLinkRate = caps->maxLinkRate;
+	}
+	
+	// All done
+	DBGLOG("igfx", "MLR: [COMM] wrapReadAUX() Maximum link rate 0x%02x has been set in the DPCD buffer.", caps->maxLinkRate);
+	return retVal;
+}
+
+IOReturn IGFX::DPCDMaxLinkRateFix::orgReadAUX(uint32_t address, void *buffer, uint32_t length) {
+	if (port != nullptr) {
+		// ICL+
+		DBGLOG("igfx", "MLR: [COMM] orgReadAUX() Routed to ICL IMP with Address = 0x%x; Length = %u.", address, length);
+		return orgICLReadAUX(port, address, buffer, length);
+	} else {
+		// CFL-
+		DBGLOG("igfx", "MLR: [COMM] orgReadAUX() Routed to CFL IMP with Address = 0x%x; Length = %u.", address, length);
+		return orgCFLReadAUX(controller, framebuffer, address, length, buffer, displayPath);
+	}
+}
+
+bool IGFX::DPCDMaxLinkRateFix::getFramebufferIndex(uint32_t &index) {
+	auto framebuffer = port != nullptr ? orgICLGetFBFromPort(*callbackIGFX->gFramebufferController, port) : this->framebuffer;
```

Comment:
Ugh, better not to alias class members. Better make it fb or something like that for the local variable.